### DISCUSSION TITLE
Why is accessing to an Eloquent magic attribute significantly slower than accessing it via `getAttributeValue` method?

### DISCUSSION THREAD

**Discussion by alexviar** [laravel/framework#45920] [Q&A]
I was profiling an API route and was surprised that accessing an Eloquent model magic attribute would take so long, then I tried using the `getAttributeValue` method directly and got significantly better performance, so much so that the profiler doesn't even consider it a hot path.
![image](https://user-images.githubusercontent.com/10967713/216329960-0528aa76-782f-4df4-937e-ec0f121d550b.png)

**Comment by alexviar** [comment]
more than 5% of execution time
![image](https://user-images.githubusercontent.com/10967713/216339721-9601e373-23d3-4dd4-82de-81ef5d460651.png)
less than 0.2% of execution time
![image](https://user-images.githubusercontent.com/10967713/216338779-2e235d28-e99b-422c-b734-8de6dd080e9d.png)

**Comment by bilalthepunjabi** [comment]
Accessing an Eloquent model's attribute through the magic method (`$model->attribute`) involves running a series of checks, such as checking if the attribute exists in the model's attributes array and running any accessors, which can add some overhead and slow down the process. On the other hand, the getAttributeValue method bypasses these checks and retrieves the value of the attribute directly from the model's attributes array, making it faster.

In general, if you need to access an attribute quickly and don't need to run any accessors or mutators, it's recommended to use the getAttributeValue method. However, if you need to run accessors or mutators, you should use the magic method instead.

**Comment by Sophist-UK** [comment]
Given this disparity in performance (at least 25x difference) and the very common usage of the magic method, it does seem to me that the code to do the magic should be looked at to see if it could be significantly optimised (either always or for the majority of calls).

**Comment by alexviar** [comment]
It was a false alarm, even if the line invoking the `getAttributeValue` method was not highlighted as a hot path, the execution times are very similar

Using magic attribute
![image](https://user-images.githubusercontent.com/10967713/216651240-2baf302a-084c-4e9b-a13a-29a188ee9d68.png)
![image](https://user-images.githubusercontent.com/10967713/216651097-1b3829f6-0b9b-4c67-8ce5-03eb06b13cea.png)

Invoking the `getAttributeValue` method directly
![image](https://user-images.githubusercontent.com/10967713/216652099-4d7a28b3-0b22-43cf-bc80-a9c7446f7e1b.png)
![image](https://user-images.githubusercontent.com/10967713/216652241-73e6cb8b-f2a3-4ff0-9916-1b2bb83e1e0a.png)

**Comment by AlliBalliBaba** [comment]
If someone is finding this later on:

To really speed up magic attributes, you'll have to add your own getter if the value is a primitive (string, integer), something like:

```php
class YourModel extends Model {

    public function getId(): ?int
    {
        return $this->attributes['id'] ?? null;
    }

}
```

Getters like this are `~50x` faster than magic attributes, so it might make sense to use it in the hot path if the value is not cast.
